### PR TITLE
Add OTLP to the specification

### PR DESCRIPTION
**Pull request by tigrannajaryan**
## Changes

We previously had OTLP definition scattered in multiple OTEPs.
This amalgamates the OTEPs and brings OTLP to the specification repo.

Related:
- [OTEP 0035](https://github.com/open-telemetry/oteps/blob/master/text/0035-opentelemetry-protocol.md) OpenTelemetry Protocol Specification
- [OTEP 0099](https://github.com/open-telemetry/oteps/blob/master/text/0099-otlp-http.md) OTLP/HTTP: HTTP Transport Extension for OTLP
- [OTEP 0122](https://github.com/open-telemetry/oteps/blob/master/text/0122-otlp-http-json.md) OTLP: JSON Encoding for OTLP/HTTP

Resolves: https://github.com/open-telemetry/opentelemetry-specification/issues/785

### CODE REVIEW

File: specification/protocol/otlp.md [hunk 1]
```diff
@@ -0,0 +1,578 @@
+# OpenTelemetry Protocol Specification
+
+OpenTelemetry Protocol (OTLP) specification describes the encoding, transport
```
Comment (SergeyKanzhelev) [inline specification/protocol/otlp.md:3]: nit: oxford comma

```suggestion
OpenTelemetry Protocol (OTLP) specification describes the encoding, transport,
```

File: specification/protocol/otlp.md [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+and delivery mechanism of telemetry data between telemetry sources, intermediate
+nodes such as collectors and telemetry backends.
```
Comment (SergeyKanzhelev) [inline specification/protocol/otlp.md:5]:
```suggestion
nodes such as collectors, and telemetry backends.
```

File: specification/protocol/otlp.md [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+
+## Table of Contents
+
```
Comment (SergeyKanzhelev) [inline specification/protocol/otlp.md:8]: typycally generated TOcs has these comments in the beggining and end for proper updates. Or tags like this:

```
<details>
<summary>
Table of Contents
</summary>
```

What tool are you using?

File: specification/protocol/otlp.md [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+  client SHOULD record the error and may retry exporting the same data. This can
+  happen when the server is temporarily unable to process the data.
+
+- Not-retryable errors indicate that processing of telemetry data failed and the
+  client MUST NOT retry sending the same telemetry data. The telemetry data MUST
+  be dropped. This can happen, for example, when the request contains bad data
+  and cannot be deserialized or otherwise processed by the server. The client
+  SHOULD maintain a counter of such dropped data.
+
+The server SHOULD indicate retryable errors using code
```
Comment (SergeyKanzhelev) [inline specification/protocol/otlp.md:169]: is it should here? There is no other way to indicate retryable error, correct? So perhaps this can be rephrased to 

```suggestion
When server chooses to indicate retryable error, it MUST be using code
```

File: specification/protocol/otlp.md [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+  st, err := status.New(codes.Unavailable, "Server is unavailable").
+    WithDetails(&errdetails.RetryInfo{RetryDelay: &duration.Duration{Seconds: 0}})
+  if err != nil {
+    log.Fatal(err)
+  }
+
+  return st.Err()
+```
+
+To indicate not-retryable errors the server is recommended to use code
```
Comment (SergeyKanzhelev) [inline specification/protocol/otlp.md:188]: Should we say that all other codes except one above are non retryable?

File: specification/protocol/otlp.md [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+
+#### Throttling
+
+OTLP allows backpressure signalling.
+
+If the server is unable to keep up with the pace of data it receives from the
+client then it SHOULD signal that fact to the client. The client MUST then
+throttle itself to avoid overwhelming the server.
+
+To signal backpressure when using gRPC transport, the server SHOULD return an
```
Comment (SergeyKanzhelev) [inline specification/protocol/otlp.md:244]: again, I think MUST is better here

File: specification/protocol/otlp.md [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+    case *errdetails.RetryInfo:
+      if t.RetryDelay.Seconds > 0 || t.RetryDelay.Nanos > 0 {
+        // Wait before retrying.
+      }
+    }
+  }
+```
+
+When the client receives this signal it SHOULD follow the recommendations
+outlined in documentation for `RetryInfo`:
```
Comment (SergeyKanzhelev) [inline specification/protocol/otlp.md:277]: Please include an url link as well

File: specification/protocol/otlp.md [hunk 8]
```diff
@@ -0,0 +1,578 @@
...
+```
+
+The value of `retry_delay` is determined by the server and is implementation
+dependant. The server SHOULD choose a `retry_delay` value that is big enough to
+give the server time to recover, yet is not too big to cause the client to drop
+data while it is throttled.
+
+#### gRPC Service and Protobuf Definitions
+
+gRPC service definitions [are here](https://github.com/open-telemetry/opentelemetry-proto/tree/master/opentelemetry/proto/collector).
```
Comment (SergeyKanzhelev) [inline specification/protocol/otlp.md:302]: Please add a phrase:

Please make sure to check the proto version and maturity level. Schemas for different signals may be at different maturity level - some stable, some in beta.

File: specification/protocol/otlp.md [hunk 9]
```diff
@@ -0,0 +1,578 @@
...
+
+On success the server MUST respond with `HTTP 200 OK`. Response body MUST be
+Protobuf-encoded
+[`ExportTraceServiceResponse`](https://github.com/open-telemetry/opentelemetry-proto/blob/e6c3c4a74d57f870a0d781bada02cb2b2c497d14/opentelemetry/proto/collector/trace/v1/trace_service.proto#L47)
+message for traces, 
+[`ExportMetricsServiceResponse`](https://github.com/open-telemetry/opentelemetry-proto/blob/e6c3c4a74d57f870a0d781bada02cb2b2c497d14/opentelemetry/proto/collector/metrics/v1/metrics_service.proto#L44)
+message for metrics and [`ExportLogsServiceResponse`](https://github.com/open-telemetry/opentelemetry-proto/blob/ca6dcbbf390d8b3ce419a931eb351d189a3eb4fa/opentelemetry/proto/collector/logs/v1/logs_service.proto#L47)
+message for logs.
+
+The server SHOULD respond with success no sooner than after successfully
```
Comment (SergeyKanzhelev) [inline specification/protocol/otlp.md:384]: Not blocking comment as this may be changed later if needed:

This effectively prevents from building udp-style http endpoints which may not be ideal. Especially for browsers when message needs to be accepted as fast as possible even if it will be dropped later. Browser or mobile app may not have another chance to send this message again.

Is it something that you think we can extend for later?

@open-telemetry/javascript-approvers what do you think here?

File: specification/protocol/otlp.md (hunk 4 above)
Comment (SergeyKanzhelev) [reply specification/protocol/otlp.md:169]: this is not blocking as it's a matter of interpretation of this phrase.

File: specification/protocol/otlp.md [hunk 10]
```diff
@@ -0,0 +1,578 @@
...
+
+OTLP is a request/response style protocols: the clients send requests, the
+server replies with corresponding responses. This document defines one requests
+and response type: `Export`.
+
+### OTLP/gRPC
+
+After establishing the underlying gRPC transport the client starts sending
+telemetry data using unary requests using `Export*Request` messages
+([`ExportTraceServiceRequest`](https://github.com/open-telemetry/opentelemetry-proto/blob/e6c3c4a74d57f870a0d781bada02cb2b2c497d14/opentelemetry/proto/collector/trace/v1/trace_service.proto#L38)
```
Comment (bogdandrutu) [inline specification/protocol/otlp.md:78]: I would avoid pining a specific commit to avoid outdated links. Maybe just refer to the proto repo in general?

File: specification/protocol/otlp.md [hunk 11]
```diff
@@ -0,0 +1,578 @@
...
+
+This specification does not use `Status.code` field and the server MAY omit
+`Status.code` field. The clients are not expected to alter their behavior based
+on `Status.code` field but MAY record it for troubleshooting purposes.
+
+The `Status.message` field SHOULD contain a developer-facing error message as
+defined in `Status` message schema.
+
+The server MAY include `Status.details` field with additional details. Read
+below about what this field can contain in each specific failure case.
```
Comment (bogdandrutu) [inline specification/protocol/otlp.md:406]: Not trying to block this, but why using the `Status` if the code is ignored. We should probably reconsider this section, an issue is good for the moment.

File: specification/protocol/otlp.md [hunk 12]
```diff
@@ -0,0 +1,578 @@
...
+  * [Multi-Destination Exporting](#multi-destination-exporting)
+* [Known Limitations](#known-limitations)
+  * [Request Acknowledgements](#request-acknowledgements)
+    * [Duplicate Data](#duplicate-data)
+  * [Partial Success](#partial-success)
+* [Future Versions and Interoperability](#future-versions-and-interoperability)
+* [Glossary](#glossary)
+* [References](#references)
+
+OTLP is a general-purpose telemetry data delivery protocol designed in the scope
```
Comment (anuraaga) [inline specification/protocol/otlp.md:38]: Comparison with prior art is nice in proposals, or a rationale doc, but I guess we don't need it in a spec that's defining the protocol? If this was a more generic list of design goals (which may require copying in some of opencensus design goals which aren't changed) it would make sense though.

File: specification/protocol/otlp.md [hunk 13]
```diff
@@ -0,0 +1,578 @@
...
+
+![Unary](img/otlp-sequential.png)
+
+Sequential operation is recommended when simplicity of implementation is
+desirable and when the client and the server are connected via very low-latency
+network, such as for example when the client is an instrumented application and
+the server is a OpenTelemetry Collector running as a local daemon (agent).
+
+The implementations that need to achieve high throughput SHOULD support
+concurrent Unary calls to achieve higher throughput. The client SHOULD send new
```
Comment (anuraaga) [inline specification/protocol/otlp.md:112]: As far as I know, for example in Java we don't ever send concurrent calls from the `BatchSpanProcessor`, but I believe that's our recommended implementation for high throughput scenarios too. Perhaps this is too much implementation detail for a spec? Just need to say implementations should support high throughput export.

File: specification/protocol/otlp.md [hunk 14]
```diff
@@ -0,0 +1,578 @@
...
+|RESOURCE_EXHAUSTED|Yes|
+|FAILED_PRECONDITION|No|
+|ABORTED|Yes|
+|OUT_OF_RANGE|Yes|
+|UNIMPLEMENTED|No|
+|INTERNAL|No|
+|UNAVAILABLE|Yes|
+|DATA_LOSS|Yes|
+
+When retrying, the client SHOULD implement an exponential backoff strategy. An
```
Comment (anuraaga) [inline specification/protocol/otlp.md:232]: Is this `MUST`? It seems weird that a client `MUST` respond to explicit throttling, but it's optional to use exponential backoff

File: specification/protocol/otlp.md [hunk 15]
```diff
@@ -0,0 +1,578 @@
...
+[JSON Mapping](https://developers.google.com/protocol-buffers/docs/proto3#json)
+for mapping between Protobuf and JSON. `trace_id` and `span_id` is base64
+encoded, not hex.
+
+#### Response
+
+Response body MUST be the appropriate serialized Protobuf message (see below for
+the specific message to use in the Success and Failure cases).
+
+The server MUST set "Content-Type: application/x-protobuf" header if the
```
Comment (anuraaga) [inline specification/protocol/otlp.md:365]: The server MUST send the same content type as it received?

File: specification/protocol/otlp.md [hunk 16]
```diff
@@ -0,0 +1,578 @@
...
+decoding and validating the request.
+
+##### Failures
+
+If the processing of the request fails the server MUST respond with appropriate
+`HTTP 4xx` or `HTTP 5xx` status code. See sections below for more details about
+specific failure cases and HTTP status codes that should be used.
+
+Response body for all `HTTP 4xx` and `HTTP 5xx` responses MUST be a
+Protobuf-encoded
```
Comment (anuraaga) [inline specification/protocol/otlp.md:394]: Maybe consider returning as response headers, similar to gRPC. Returning a binary payload for an error response is the easiest way to lose SRE friends :)

File: specification/protocol/otlp.md [hunk 17]
```diff
@@ -0,0 +1,578 @@
...
+#### Default Port
+
+The default network port for OTLP/HTTP is 55681. There is currently an [open
+issue](https://github.com/open-telemetry/opentelemetry-collector/issues/1256) to
+use the same port for OTLP/gRPC and OTLP/HTTP. In that case this spec will be
+updated to use the same default port for OTLP/gRPC and OTLP/HTTP.
+
+## Implementation Recommendations
+
+### Multi-Destination Exporting
```
Comment (anuraaga) [inline specification/protocol/otlp.md:474]: Is this related to OTLP?

File: specification/protocol/otlp.md [hunk 18]
```diff
@@ -0,0 +1,578 @@
...
+
+OTLP will evolve and change over time. Future versions of OTLP must be designed
+and implemented in a way that ensures that clients and servers that implement
+different versions of OTLP can interoperate and exchange telemetry data. Old
+clients must be able to talk to new servers and vice versa. If new versions of
+OTLP introduce new functionality that cannot be understood and supported by
+nodes implementing the old versions of OTLP the protocol must regress to the
+lowest common denominator from functional perspective.
+
+When possible the interoperability SHOULD be ensured between all versions of
```
Comment (anuraaga) [inline specification/protocol/otlp.md:522]: I'd say MUST, backwards compatibility is really important. If a completely unavoidable issue happens (e.g., security hole), I'd generally expect all the old versions to be obsolete anyways.

File: specification/protocol/otlp.md [hunk 19]
```diff
@@ -0,0 +1,578 @@
...
+   which define the new capabilities and typically can be implemented by making
+   a discovery request/response message exchange from the client to server. The
+   mandatory capabilities defined by this specification are implied and do not
+   require a discovery. The implementation which supports a new, optional
+   capability MUST adjust its behavior to match the expectation of a peer that
+   does not support a particular capability.
+
+The current version of OTLP is the initial version that describes mandatory
+capabilities only. Implementations of this specification SHOULD NOT attempt to
+detect the capabilities of their peers and should operate as defined in this
```
Comment (anuraaga) [inline specification/protocol/otlp.md:555]: I'm not sure where to look for "as defined in this document" since we don't talk about business logic. Is this paragraph needed?

File: specification/protocol/otlp.md (hunk 3 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md:8]: Markdown All in One extension for VS Code.

File: specification/protocol/otlp.md (hunk 9 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md:384]: >udp-style http endpoints

I am not familiar with the concept. Can you please clarify?

File: specification/protocol/otlp.md (hunk 1 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md]: Fixed.

File: specification/protocol/otlp.md (hunk 7 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md]: Done.

File: specification/protocol/otlp.md (hunk 8 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md]: Added.

File: specification/protocol/otlp.md (hunk 12 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md:38]: I agree. Removed.

File: specification/protocol/otlp.md (hunk 15 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md:353]: Good point. Added.

File: specification/protocol/otlp.md (hunk 2 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md:5]: I may be wrong but I think it is not needed. From https://www.grammarly.com/blog/comma/
>Don’t separate two nouns that appear together as a compound subject or compound object.
>...you don’t need a comma when you’re simply listing two items.

File: specification/protocol/otlp.md (hunk 10 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md]: Good point. Added one link to the proto repo and removed the common-specific links.

File: specification/protocol/otlp.md (hunk 14 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md:208]: The client may decide to implement a different backoff strategy for whatever reason. SHOULD is typically used instead of MUST when deviating from the recommendation is undesirable but is not catastrophic. In this case it is not catastrophic to use a different backoff policy.

Comment (tigrannajaryan) [reply specification/protocol/otlp.md:208]: Perhaps the same argument can be made about the response to throttling and it is also better to use SHOULD instead of MUST for that as well, but I think it is slightly more important to mandate that the client must honor server's desire to back off, and a bit less important to specify how exactly it should be honored.

Comment (tigrannajaryan) [reply specification/protocol/otlp.md:208]: Either way, I think the difference in this particular case is small, happy to change if you have a strong opinion about it.

File: specification/protocol/otlp.md (hunk 16 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md:370]: Do you suggest to return the exact same data encoded in a more human-readable format in the response headers (i.e. Protobuf-JSON)? Or you suggest to include a human-readable text in a response header _in addition_ to the response body?

We could include a copy of `Status.message` (which is supposed to be human-readable) in a response header, but I think we still need to return the full `Status` (which may contain `details`) in the body.

File: specification/protocol/otlp.md (hunk 11 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md:382]: Two reasons:
1. To keep the response more uniform with OTLP/gRPC.
2. To allow to include `Status.details`. This can contain important information for BadRequest response.

File: specification/protocol/otlp.md (hunk 18 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md]: Done.

File: specification/protocol/otlp.md (hunk 6 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md]: Done.

File: specification/protocol/otlp.md (hunk 4 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md]: Changed to MUST.

File: specification/protocol/otlp.md (hunk 13 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md:91]: I think it is important to provide a bit more guidance and explain why. If one is not an expert in protocols and their implementations it is not immediately obvious that "high throughput" coupled with half-duplex protocol leads to a requirement to do concurrent requests.

File: specification/protocol/otlp.md (hunk 17 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md:450]: Strictly speaking no, it is applicable to any exporting protocol, but still important implementation detail when exporting to multiple destinations. If there is a better place for this information I can move it.

File: specification/protocol/otlp.md (hunk 19 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md:531]: >I'm not sure where to look for "as defined in this document" 

"this document" is `otlp.md`, the document that one is reading. I can try to rephrase if it is not clear.

>since we don't talk about business logic.

I am confused by the reference to "business logic" and I don't understand what the problem is. 

>Is this paragraph needed?

Yes, I think it is important. In other words it says: "don't try to be clever and implement only parts of this spec and try to detect which parts of the spec your peers implemented. This is the first version, implement it fully as described".

File: specification/protocol/otlp.md (hunk 17 above)
Comment (anuraaga) [reply specification/protocol/otlp.md:450]: I think it could make sense closer to `Span Processor` or `Exporter` in this file https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md

File: specification/protocol/otlp.md (hunk 19 above)
Comment (anuraaga) [reply specification/protocol/otlp.md:531]: I think the capabilities mentioned here are for example, the capability to send a `Status`. So it seems like this paragraph is talking more about business logic, how it handles `Status`, rather than the protocol itself. Another way to put it, I think even if an implementation did attempt to detect capabilities of their peers, I don't think that would conflict with anything else defined in this document. So I found it confusing what this is referring to. Maybe not `defined in this document` but `defined in this proto` (link to proto)?

File: specification/protocol/otlp.md (hunk 9 above)
Comment (SergeyKanzhelev) [reply specification/protocol/otlp.md:360]: I mean when http endpoint just takes the payload and replies 200 OK right away. So browser is not waiting on deserialization.

Comment (tigrannajaryan) [reply specification/protocol/otlp.md:360]: I understand. If there are good reasons to do that then it is indeed possible to do that. This clause is marked "SHOULD" which allows the implementor to deviate if they have good reasons to. Do you think "SHOULD" is too strong here?
(For the sake of making a progress we can probably merge as is, since this PR is merely bringing what already was approves as an OTEP to this repository and I can submit an issue to look into this later).

File: specification/protocol/otlp.md (hunk 14 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md:208]: Sounds good, resolving this comment.

File: specification/protocol/otlp.md (hunk 17 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md:450]: It is applicable not only to SDK but any exporting implementation, for example to the Collector, so I am not sure that would be the right place. I can delete from here for now if it looks misplaced.

File: specification/protocol/otlp.md (hunk 19 above)
Comment (tigrannajaryan) [reply specification/protocol/otlp.md]: I re-read the paragraph and I think you are right, it does not add value to this document. Deleted.

**Comment by tigrannajaryan** [comment]
@open-telemetry/specs-approvers please review.

File: specification/protocol/otlp.md (hunk 17 above)
Comment (anuraaga) [reply specification/protocol/otlp.md:450]: To be honest, it's always hard to talk about "exporter" to team members since they get confused between SDK and collector. I consider them the same, either you export directly or you proxy through the collector. So indeed, SDK does seem like a bad file for this and other aspects of export, in another PR we should probably create a separate section for export specifically, as it stands for both SDK and collector

File: specification/protocol/otlp.md (hunk 16 above)
Comment (anuraaga) [reply specification/protocol/otlp.md:370]: Not going to block on this but still strongly recommend returning text somewhere for errors, headers or payload or otherwise. Protobuf already makes SREs hate gRPC in my experience, if we remove error headers too they'll really be unhappy.

Comment (tigrannajaryan) [reply specification/protocol/otlp.md:370]: Added an issue to address later: https://github.com/open-telemetry/opentelemetry-specification/issues/808

**Comment by SergeyKanzhelev** [comment]
@anuraaga please file issues for the concerns you raised if you believe those needs to be addressed before GA.

As for this PR - it seems to be OK from number of approvals and cool off period. Also OTEPs it is created from was already discussed and agreed on. Merging.
